### PR TITLE
In Markdown table cells, apply HTML escaping only to code blocks, and apply it properly

### PR DESCRIPTION
**Pull request by tetromino**
In Markdown table cells, apply HTML escaping only to code blocks, and apply it properly

Since #161 removed HTML escaping for defaults and function docstrings,
we should do the same for attribute and param docs in table cells.

The only limitations Markdown places on table cells are:
* no pipe characters (they must be escaped with a backslash)
* no newlines (they must be transformed into `<br>` or an HTML entity)

The latter restriction makes it impossible to have a fenced code block inside a table cell.

Therefore:
* we do not escape HTML or Markdown markup outside a fenced code block
* we keep existing logic for escaping newlines outside a fenced code block
* we fix fence detection (e.g. allowing more than 3 fence characters to support embedded code blocks in code blocks, allowing tildes as fence characters, properly handling language names, etc.);
* in code block content, we escape HTML, and we escape newlines as HTML entities (since `<br>` does not work in a `<pre><code>` block) - finally fixing code block newlines in table cells.
    
This is a followup to #161.

Partially addresses #118

### CODE REVIEW

**Comment by tetromino** [comment]
This is a draft PR to be merged after #166 lands

@fmeum fyi

**Comment by tetromino** [comment]
I have refactored markdownCellFormat to make it feasible to extend it later for formatting lists etc. inside markdown tables.

File: src/main/java/com/google/devtools/build/skydoc/rendering/MarkdownUtil.java [hunk 1]
```diff
@@ -46,55 +47,131 @@ public MarkdownUtil(String extensionBzlFile) {
...
-    String resultString = htmlEscape(docString.trim());
+  public static String markdownCellFormat(String docString) {
+    return new MarkdownCellFormatter(docString).format();
+  }
 
-    resultString = replaceWithTag(resultString, "```", "<pre><code>", "</code></pre>");
-    resultString = replaceWithTag(resultString, "`", "<code>", "</code>");
+  // See https://github.github.com/gfm
+  private static final class MarkdownCellFormatter {
+    private final ImmutableList<String> lines;
```
Comment (brandjon) [inline src/main/java/com/google/devtools/build/skydoc/rendering/MarkdownUtil.java:71]: Please add one-line comments to doc the invariant for these fields. (Especially that currentLine is 0-indexed.)

File: src/main/java/com/google/devtools/build/skydoc/rendering/MarkdownUtil.java [hunk 2]
```diff
@@ -46,55 +47,131 @@ public MarkdownUtil(String extensionBzlFile) {
...
 
-    boolean openTagNext = true;
-    int index = remainingString.indexOf(stringToReplace);
-    while (index > -1) {
-      resultString.append(remainingString, 0, index);
-      resultString.append(openTagNext ? openTag : closeTag);
-      openTagNext = !openTagNext;
-      remainingString = remainingString.substring(index + stringToReplace.length());
-      index = remainingString.indexOf(stringToReplace);
+    String format() {
```
Comment (brandjon) [inline src/main/java/com/google/devtools/build/skydoc/rendering/MarkdownUtil.java:84]: Nit: Javadoc that this consumes the whole input and yields the result.

File: src/main/java/com/google/devtools/build/skydoc/rendering/MarkdownUtil.java [hunk 3]
```diff
@@ -46,55 +47,131 @@ public MarkdownUtil(String extensionBzlFile) {
...
+    }
+
+    /**
+     * If blank lines appear at {@link #currentLine}, render to {@link #result}, update {@link
+     * #currentLine} to point to end of break, and return true.
+     */
+    private boolean formatParagraphBreak() {
+      int numEmptyLines = 0;
+      for (int i = currentLine; i < lines.size(); i++) {
+        if (lines.get(i).isEmpty() || lines.get(i).equals("\r")) {
```
Comment (brandjon) [inline src/main/java/com/google/devtools/build/skydoc/rendering/MarkdownUtil.java:146]: I'm not sure if the \r can organically occur, absent user-written escapes in string literals (which break docstring parsing assumptions in other ways). The lexer normalizes \r\n into \n. You can leave it if you're not sure if it's needed.

File: src/main/java/com/google/devtools/build/skydoc/rendering/MarkdownUtil.java [hunk 4]
```diff
@@ -46,55 +47,131 @@ public MarkdownUtil(String extensionBzlFile) {
...
+          return true;
+        }
+      }
+      // We did not find the closing fence.
+      return false;
+    }
+
+    /**
+     * If blank lines appear at {@link #currentLine}, render to {@link #result}, update {@link
+     * #currentLine} to point to end of break, and return true.
```
Comment (brandjon) [inline src/main/java/com/google/devtools/build/skydoc/rendering/MarkdownUtil.java:141]: "to point to the last line of the break"

**Review by brandjon — APPROVED** [review]
I left some comments on the parsing logic but didn't go through the whole algorithm with a fine-toothed comb.

File: src/main/java/com/google/devtools/build/skydoc/rendering/MarkdownUtil.java [hunk 5]
```diff
@@ -46,55 +47,152 @@ public MarkdownUtil(String extensionBzlFile) {
...
    * <p>For example: 'Information with <brackets>.' becomes 'Information with &lt;brackets&gt;'.
    */
-  public String htmlEscape(String docString) {
+  public static String htmlEscape(String docString) {
     return docString.replace("<", "&lt;").replace(">", "&gt;");
   }
 
+  /** Returns a string that escapes newlines and carriage returns with HTML entities. */
+  private static String newlineEscape(String docString) {
+    return docString.replace("\n", "&#10;").replace("\r", "&#13;");
```
[the remark this answers is not in the document]
Comment (tetromino) [reply src/main/java/com/google/devtools/build/skydoc/rendering/MarkdownUtil.java:172]: A backslash-escaped backslash would be `"\\n"` and we would not replace it. (Note that `"\n"` is the Java string value containing the backslash character, not the backslash escape.)

I have added a test to verify we are doing the right thing here.

File: src/main/java/com/google/devtools/build/skydoc/rendering/MarkdownUtil.java (hunk 2 above)
Comment (tetromino) [reply src/main/java/com/google/devtools/build/skydoc/rendering/MarkdownUtil.java:88]: Fixed, good point.

File: src/main/java/com/google/devtools/build/skydoc/rendering/MarkdownUtil.java (hunk 4 above)
Comment (tetromino) [reply src/main/java/com/google/devtools/build/skydoc/rendering/MarkdownUtil.java]: Fixed.

File: src/main/java/com/google/devtools/build/skydoc/rendering/MarkdownUtil.java (hunk 3 above)
Comment (tetromino) [reply src/main/java/com/google/devtools/build/skydoc/rendering/MarkdownUtil.java]: Fixed. There are several points where \r\n would be stripped - in the lexer, in the doc string dedenter, and if someone would add \r\n to the proto manually, we strip it via String.lines()

**Pull request merged**